include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,428 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_ASYNC_POOL_H_
+#define DALI_CORE_MM_ASYNC_POOL_H_
+
+#include <algorithm>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+#include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/free_list.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/cuda_event_pool.h"
+#include "dali/core/cuda_stream.h"
+
+namespace dali {
+namespace mm {
+
+template <memory_kind kind, class FreeList, class LockType, class Upstream = memory_resource<kind>>
+class async_pool_base : public stream_aware_memory_resource<kind> {
+ public:
+  explicit async_pool_base(Upstream *upstream) : global_pool_(upstream) {
+  }
+  ~async_pool_base() {
+    try {
+      synchronize();
+    } catch (const CUDAError &e) {
+      if ((e.is_rt_api() && e.rt_error() != cudaErrorCudartUnloading) ||
+          (e.is_drv_api() && e.drv_error() != CUDA_ERROR_DEINITIALIZED))
+        std::terminate();
+    }
+  }
+
+  /**
+   * @brief Waits until all pending frees are finished.
+   */
+  void synchronize() {
+    {
+      std::lock_guard<LockType> guard(lock_);
+      for (auto &kv : stream_free_) {
+        if (!kv.second.free_list.head)
+          continue;
+        if (!sync_stream_)
+          sync_stream_ = CUDAStream::Create(true);
+        CUDA_DTOR_CALL(cudaStreamWaitEvent(sync_stream_, kv.second.free_list.head->event, 0));
+      }
+    }
+    CUDA_DTOR_CALL(cudaStreamSynchronize(sync_stream_));
+  }
+
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    return allocate_from_global_pool(bytes, alignment);
+  }
+
+  void do_deallocate(void *mem, size_t bytes, size_t alignment) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    char *ptr = static_cast<char *>(mem);
+    restore_original_params(ptr, bytes, alignment);
+    return global_pool_.deallocate(ptr, bytes, alignment);
+  }
+
+  /**
+   * @brief Tries to recycle per-stream free memory or allocates from the global pool.
+   *
+   * There are four cases:
+   * (1) Attempt to allocate from stream-specific free list - a smallest suitable block is used,
+   *     padding is applied, if necessary, and recorded.
+   * (2) If (1) fails and there are no pending per-stream free operations, the requested memory
+   *     is allocated from the global pool.
+   * (3) If (1) fails and there are free pending per-stream operations, an attempt is made to get
+   *     memory from the free list of the underlying global pool, without using upstream resource.
+   *     If this allocation succeeds, the memory allocated this way is returned.
+   * (4) If (3) fails, the pending per-stream frees are collected to the global pool and then
+   *     an allocation is made from the global pool.
```

Comment:
Do I map the cases correctly:
1) it != stream_free_.end() && ptr
2) it == stream_free_.end()
3) it != stream_free_.end() && !ptr
4) ?

Reply:
Or the description includes `allocate_from_global_pool`?
If so I would clarify it a bit.

Reply:
Yes, it includes `allocate_from_global_pool` - it was factored out after this comment had been written.

Reply:
Can you update it then?